src/kafka/producer/sender.cc

```diff
@@ -0,0 +1,261 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Copyright (C) 2019 ScyllaDB Ltd.
+ */
+
+#include "../protocol/produce_request.hh"
+#include <seastar/core/future-util.hh>
+#include "sender.hh"
+
+namespace seastar {
+
+namespace kafka {
+
+sender::sender(lw_shared_ptr<connection_manager> connection_manager,
+        lw_shared_ptr<metadata_manager> metadata_manager)
+            : _connection_manager(std::move(connection_manager)), _metadata_manager(std::move(metadata_manager)) {}
+
+sender::connection_id sender::broker_for_topic_partition(const std::string& topic, int32_t partition_index) {
+    for (const auto& current_topic : *_metadata_manager->get_metadata()._topics) {
+        if (*current_topic._name != topic) {
+            continue;
+        }
+        for (const auto& current_partition : *current_topic._partitions) {
+            if (*current_partition._partition_index == partition_index) {
+                return broker_for_id(*current_partition._leader_id);
+            }
+        }
+    }
+    return {};
+}
+
+bool sender::is_broker_for_topic_partition_known(const std::string &topic, int32_t partition_index) {
+    for (const auto& current_topic : *_metadata_manager->get_metadata()._topics) {
+        if (*current_topic._name != topic) {
+            continue;
+        }
+        if (current_topic._error_code != error::kafka_error_code::NONE) {
+            continue;
+        }
+        for (const auto& current_partition : *current_topic._partitions) {
+            if (*current_partition._partition_index == partition_index
+                && current_partition._error_code == error::kafka_error_code::NONE) {
+                return true;
+            }
+        }
+    }
```

Comment:
The code for these two methods is almost the same. Isn't there a way to merge them into one? Especially given, that the only difference is the one `if` statement with `kafka_error_code`.

Reply:
It should probably be moved to `metadata_response` and also optimized (sort topics, partitions and search for them in `log N` not linearly).

Reply:
I'm not requiring updating the time complexity as of now (although leave a TODO in the code), but I'd say it's better to have this code better organized right now than submit nearly duplicated methods.